File: frontend/src/order/object/core/OrderItemInput.ts
import type { MenuItemId, Quantity } from '@/shared/object/domain/DomainObjects'
import type { OrderItemSelection } from '@/order/object/core/OrderItemSelection'

export type OrderItemInput = {
  menuItemId: MenuItemId
  quantity: Quantity
  selections: OrderItemSelection[]
}

File: backend/src/main/scala/domain/order/OrderItemInput.scala
package domain.order

import domain.shared.given

import io.circe.{Decoder, Encoder}
import io.circe.generic.semiauto.*
import domain.shared.*

final case class OrderItemInput(
    menuItemId: MenuItemId,
    quantity: Quantity,
    selections: List[OrderItemSelection],
)
object OrderItemInput:
  given Encoder[OrderItemInput] = deriveEncoder
  given Decoder[OrderItemInput] = Decoder.instance { cursor =>
    for
      menuItemId <- cursor.get[MenuItemId]("menuItemId")
      quantity <- cursor.get[Quantity]("quantity")
      selections <- cursor.getOrElse[List[OrderItemSelection]]("selections")(List.empty)
    yield OrderItemInput(menuItemId = menuItemId, quantity = quantity, selections = selections)
  }
